apisix/conf_server.lua

```diff
@@ -136,13 +138,25 @@ local function pick_node(ctx)
 
         ctx.server_picker = server_picker
         ctx.balancer_server = server
-        res = create_resolved_result(server)
+
+        for _, r in ipairs(resolved_results) do
+            if r.server == server then
+                res = r
+                break
+            end
+        end
     else
         res = resolved_results[1]
     end
 
     ctx.balancer_ip = res.host
     ctx.balancer_port = res.port
+
+    ngx_var.upstream_host = res.domain or res.host
+    if balancer.recreate_request and ngx.get_phase() == "balancer" then
+        balancer.recreate_request()
+    end
```

Comment:
Why `recreate_request `?

Reply:
So we can use a new Host, see https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/balancer.md#recreate_request